feat: kpi forecasting add funnel_forecast unit tests

jaredsnyder opened the pull request:
Adds unit tests for funnel_forecast and does some light refactoring

Checklist for reviewer:

- [ ] Commits should reference a bug or github issue, if relevant (if a bug is
  referenced, the pull request should include the bug number in the title)
- [ ] Scan the PR and verify that no changes (particularly to
  `.circleci/config.yml`) will cause environment variables (particularly
  credentials) to be exposed in test logs
- [ ] Ensure the container image will be using permissions granted to
  [telemetry-airflow](https://github.com/mozilla/telemetry-airflow/)
  responsibly.

jaredsnyder commented:
Made a notebook to compare outputs.  For most of the rows (where the optimization lands on the same parameters), the difference is within 1% but there are some significant outliers.  Not sure if this is inevitable due to the stochastic nature of the prediction or not.  Notably we didn't have this issue with `prophet_forecast`

https://colab.research.google.com/drive/1NxD5eVwZ0Vw3UhZtZoIev5Q44qEEjSTJ#scrollTo=CjguJpSKsv7k

jaredsnyder commented:
The seed is set in both branches, so the results should be the same.  Investigating

jaredsnyder commented:
Validation notebook is updated and all tests pass.  Two changes were necessary:

- A bug we fixed was switching the values in `_add_regressor` so 1 was assigned when true and 0 when False, while in main it is the other way around.  This did change the values, usually by small amounts but occasionally by more.  It may be worth revisiting whether or not this change is worth it.  To check, I reverted this change to produce the `_branch_0s` tables
- there was another bug involving using the aggregate function on strings.  I updated the function and added some tests.

NB: [../eam-integrations/scripts/workday_everfi_integration.py](https://colab.research.google.com/drive/1NxD5eVwZ0Vw3UhZtZoIev5Q44qEEjSTJ#scrollTo=68HyVEYZ5C1Y)

m-d-bowerman commented on jobs/kpi-forecasting/kpi_forecasting/models/funnel_forecast.py:
```diff
@@ -311,58 +389,53 @@ def _auto_tuning(self, segment_settings: SegmentModelSettings) -> Dict[str, floa
 
-    def _predict(self, segment_settings: SegmentModelSettings) -> pd.DataFrame:
+    def _predict(
+        self, dates_to_predict_raw: pd.DataFrame, segment_settings: SegmentModelSettings
```
Was the purpose of passing the `dates_to_predict` df as an attribute to make testing simpler?

jaredsnyder replied:
Yeah that's the main reason

m-d-bowerman commented on jobs/kpi-forecasting/kpi_forecasting/pandas_extras.py:
```diff
@@ -26,4 +26,32 @@ def aggregate_to_period(
+    x_string = x.select_dtypes(include=["datetime64", object])
+    x_numeric = x.select_dtypes(include=["float", "int", "datetime64"])
+
+    if set(x_string.columns) | set(x_numeric.columns) != set(x.columns):
```
Can you explain why this check is necessary? I can't think of an instance in either `ProphetForecast` or `FunnelForecast` that might cause this check to fail.

jaredsnyder replied:
It's a fail-safe.  The idea is that this function doesn't cover all types so if an expected one appears it will throw an error instead of quietly doing something weird (There whole motivation for adding tests for this function was it essentially "quietly did something weird*" when I updated the code).  IE if a new column was boolean valued this would throw an error.

* The string-valued columns got aggregated like `+` so you'd see values like `desktopdesktopdesktopdesktopdesktop`, it was funny

m-d-bowerman commented on jobs/kpi-forecasting/kpi_forecasting/pandas_extras.py:
```diff
@@ -26,4 +26,32 @@ def aggregate_to_period(
+    # all values of x_string should be the same because it is just the dimensions
+    x_string_agg = x_string.drop_duplicates().reset_index(drop=True)
+
+    if len(x_string_agg) != len(x_numeric_agg):
```
This checks that there is a single value on each aggregation date for each combination of segment dimensions, correct?

jaredsnyder replied:
Right, I'm assuming that the string values are one-to-one with dimensions so they're unique and this essentially checks that and raises and error if that's not the case

m-d-bowerman commented on jobs/kpi-forecasting/kpi_forecasting/tests/test_funnel_forecast.py:
```diff
@@ -0,0 +1,1603 @@
+        for el in output_df.columns
+        if el not in ["value", "value_low", "value_mid", "value_high"]
+    ]
+    assert not pd.isna(output_df[non_metric_columns]).any(axis=None)
```
Why filter to `non_metric_columns` here? I don't think there should be null values in any of the columns.

jaredsnyder replied:
Currently for the observed data the percentile columns ("value_low", "value_mid", "value_high") are NaN.  Is that expected?

m-d-bowerman replied:
Ahh, you right, whoops

m-d-bowerman reviewed (approved):
The changes to `funnel_forecast` look good. The test cases, too, look sufficient to catch errors in the methods that might arise from future changes to the classes. I left a couple of comments for my clarity, but let's go ahead with it rather than having this persist as a blocker.

m-d-bowerman commented:
> [Validation NB](https://colab.research.google.com/drive/1NxD5eVwZ0Vw3UhZtZoIev5Q44qEEjSTJ#scrollTo=68HyVEYZ5C1Y) is updated and all tests pass. Two changes were necessary:
> 
>     * A bug we fixed was switching the values in `_add_regressor` so 1 was assigned when true and 0 when False, while in main it is the other way around.  This did change the values, usually by small amounts but occasionally by more.  It may be worth revisiting whether or not this change is worth it.  To check, I reverted this change to produce the `_branch_0s` tables
> 
>     * there was another bug involving using the aggregate function on strings.  I updated the function and added some tests.

I added a bit to the notebook, checking for differences between values, and all differences were 0. The lil bit is right before the components section; have a look, make sure I didn't goof up the merge or sets I used?

jaredsnyder commented:
> > [Validation NB](https://colab.research.google.com/drive/1NxD5eVwZ0Vw3UhZtZoIev5Q44qEEjSTJ#scrollTo=68HyVEYZ5C1Y) is updated and all tests pass. Two changes were necessary:
> > ```
> > * A bug we fixed was switching the values in `_add_regressor` so 1 was assigned when true and 0 when False, while in main it is the other way around.  This did change the values, usually by small amounts but occasionally by more.  It may be worth revisiting whether or not this change is worth it.  To check, I reverted this change to produce the `_branch_0s` tables
> > 
> > * there was another bug involving using the aggregate function on strings.  I updated the function and added some tests.
> > ```
> 
> @jaredsnyder I added a bit to the notebook, checking for differences between values, and all differences were 0. The lil bit is right before the components section; have a look, make sure I didn't goof up the merge or sets I used?

@m-d-bowerman : that's because the tables you were looking at ran the data with that change reverted.  I loaded the tables with the change and made some plots to compare the differences here https://colab.research.google.com/drive/1NxD5eVwZ0Vw3UhZtZoIev5Q44qEEjSTJ#scrollTo=Look_at_distribution_of_differences_when_changing_1_to_0

m-d-bowerman commented:
Updates look good :shipit:  🦜
